data-prepper-plugins/grok-prepper/README.md

```diff
@@ -1,59 +1,241 @@
+can be extremely helpful.
+
+## Basic Grok Usage
+
+To get started with grok using Data Prepper, create the following `pipeline.yaml`.
+```yaml
+grok-pipeline:
+  source:
+    file:
+      path: "/full/path/to/grok_logs_json.log"
+  prepper:
+    - grok:
+        match:
+          message: ['%{IPORHOST:clientip} \[%{HTTPDATE:timestamp}\] %{NUMBER:response_status:int}']
+  sink:
+    - stdout:
+```
+
+Create a file named `grok_logs_json.log` and replace the `path` in the file source of your `pipeline.yaml` with the path of this file.
+
+The grok configuration from the `pipeline.yaml` will match the value in the `message` key of each log for a pattern matching `%{IPORHOST:clientip} \[%{HTTPDATE:timestamp}\] %{NUMBER:response_status:int}`.
+These three patterns (`IPORHOST`, `HTTPDATE`, and `NUMBER`) are default patterns.
 
-Example `.yaml` configuration
+Copy the following log into `grok_logs_json.log` in order to align with this.
 
 ```
-prepper:
-  - grok:
-     match:
-       message: [ "%{COMMONAPACHELOG}" ]
+{"message": "127.0.0.1 198.126.12 [10/Oct/2000:13:55:36 -0700] 200"}
 ```
 
-## Configuration
+When you run Data Prepper with this `pipeline.yaml` passed in, you should see the following standard output.
+
+```
+{ 
+  "message":"127.0.0.1 198.126.12 [10/Oct/2000:13:55:36 -0700] 200",
+  "response_status":200,
+  "clientip":"198.126.12",
+  "timestamp":"10/Oct/2000:13:55:36 -0700"
+}
+```
 
-TODO: provide examples for using each configuration
+As you can see, the extra keys for `clientip`, `timestamp`, and `response_status` have been pulled out from the original message.
 
+## Configuration
 * `match` (Optional): A `Map<String, List<String>>` that specifies which keys of a Record to match which patterns against. Default value is `{}`
 
+This example match configuration will check logs for a `message` key, and if it exists, will match the value in this `message` key first against the `SYSLOGBASE` pattern, and then against the `COMMONAPACHELOG` pattern.
+It will then check logs for a `timestamp` key, and if it exists, will attempt to match the value in this `timestamp` key against the `TIMESTAMP_ISO8601` pattern.
 
-* `keep_empty_captures` (Optional): A `boolean` that specifies whether `null` captures should be kept. Default value is `false`
-  
-  
+Note that by default, matching will be done until there is a successful match. So if there is a successful match against the value in the `message` key for a pattern of `SYSLOGBASE`, no attempted matching will be done 
+for either the `COMMONAPACHELOG` or `TIMESTAMP_ISO8601` pattern. If you would like to match logs against every pattern in `match` no matter what, then see [break_on_match](#break_on_match).
+
+```yaml
+prepper:
+  - grok:
+      match:
+        message: ['%{SYSLOGBASE}', "%{COMMONAPACHELOG}"]
+        timestamp: ["%{TIMESTAMP_ISO8601}"]  
+```
+
+&nbsp;
```

Comment:
I think you can remove this since it's adding extra new lines. There's some more places down below which I will point out.

Reply:
This was intentional on my part. Without them I thought everything looked too cluttered with the final rendering of the README.

Reply:
Gotcha, I actually took a look at your README after the markdown is applied, see [here](https://github.com/opensearch-project/data-prepper/blob/c7dd20b255d34d27dd9a8bd12431aa50be19902b/data-prepper-plugins/grok-prepper/README.md) and the spaces stood out to me then. But this is isn't blocking an approval.

Reply:
I can get rid of them if you think it looks better without them. But some of these configuration sections are so long that I didn't want them to blend in with some of the others.